BuildForSDG/SmeVest-Backend
# Ft confirm email

## Description
- Created validator for  confirming email,
- Created confirm email method in the auth controller
- Wrote unit test for validator,  and controller
- Created the auth middleware
- Created the confirm email temaplate and its functionality using nodemailer
Finishes #8
## How Has This Been Tested?
1. In your terminal, run `git clone https://github.com/BuildForSDG/SmeVest-Backend.git`
2. Run `cd SmeVest-Backend`
3. Run `git fetch origin ft-confirm-email`
4. Run `git checkout origin/ft-confirm-email`
5. Run `yarn install`
6. create a `.env` file in the root directory then add details as seen in `.env-example` file
7. Run `yarn run test` to run unit tests
8. Run `yarn run start:dev` to start the development server
7. To test with postman use `http://localhost:4000/api/v1/auth/signup` to sign up a new user with appropriate parameters, **If you sign up with a valid mail you will get a confirm email message in your inbox containing your email confirm code**

- Sample request:
```json
{
	"email": "emails@mail.com",
	"password": "password",
        "role": "sme"
}
```

- Sample success response :
 ```json
{
    "status": 201,
    "message": "User Created Successfully",
    "data": {
        "isVerified": false,
        "_id": "5ebae6400e9af8219ca81ff1",
        "email": "emails@mail.com",
        "password": "$2b$10$CNSmyli5fBVwUT26S95GieO1mNd5QlzVYRpi9N9yb5OOvF/8S.Rfq",
        "role": "sme",
        "createdAt": "2020-05-12T18:09:04.141Z",
        "emailConfirmCode": "GnCPH0buwI4qtNZGM6JIoOkeAlRG3f8b",
        "__v": 0
    },
    "token": "eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpZCI6IjVlYmFlNjQwMGU5YWY4MjE5Y2E4MWZmMSIsImlhdCI6MTU4OTMwNjk0NH0.yxypHlC2sDi6yb7clX1eb3pv0qFZoyms3WtrO4pIbEg"
}
```
7. Then copy the **emailConfirmCode** then use `http://localhost:4000/api/v1/auth/emails/confirm` to confirm a new user with appropriate parameters

- Sample request:
```json
{
	"token": "GnCPH0buwI4qtNZGM6JIoOkeAlRG3f8b",
}
```

- Sample success response :
 ```json
{
    "status": 200,
    "message": "Email confirmed successfully",
    "data": {
        "isVerified": true,
        "_id": "5ec4453953ef6924585e79b7",
        "email": "emails@mail.com",
        "role": "sme",
        "createdAt": "2020-05-19T20:44:41.851Z",
        "emailConfirmCode": null,
        "__v": 0,
        "emailConfirmedAt": "2020-05-12T18:09:04.141Z",
        "token": "eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpZCI6IjVlYzQ0NTM5NTNlZjY5MjQ1ODVlNzliNyIsImlhdCI6MTU4OTkyMTEwNX0.2YCye4NyILA_EfbFv3kifmff51UveCNSZMDgq8-lkU0"
    }
}
```

## Checklist:
